service/src/main/kotlin/no/nav/su/se/bakover/service/revurdering/RevurderingServiceImpl.kt

```diff
@@ -0,0 +1,175 @@
+import no.nav.su.se.bakover.domain.behandling.BeregnetRevurdering
+import no.nav.su.se.bakover.domain.behandling.OpprettetRevurdering
+import no.nav.su.se.bakover.domain.behandling.Revurdering
+import no.nav.su.se.bakover.domain.behandling.RevurderingTilAttestering
+import no.nav.su.se.bakover.domain.behandling.SimulertRevurdering
+import no.nav.su.se.bakover.domain.beregning.Beregning
+import no.nav.su.se.bakover.domain.beregning.fradrag.Fradrag
+import no.nav.su.se.bakover.domain.brev.LagBrevRequest
+import no.nav.su.se.bakover.domain.oppgave.OppgaveConfig
+import no.nav.su.se.bakover.service.brev.BrevService
+import no.nav.su.se.bakover.service.oppgave.OppgaveService
+import no.nav.su.se.bakover.service.person.PersonService
+import no.nav.su.se.bakover.service.sak.SakService
+import no.nav.su.se.bakover.service.utbetaling.UtbetalingService
+import java.util.UUID
+
+internal class RevurderingServiceImpl(
+    private val sakService: SakService,
+    private val utbetalingService: UtbetalingService,
+    private val revurderingRepo: RevurderingRepo,
+    private val oppgaveService: OppgaveService,
+    private val personService: PersonService,
+    private val microsoftGraphApiClient: MicrosoftGraphApiOppslag,
+    private val brevService: BrevService,
+) : RevurderingService {
+
+    override fun opprettRevurdering(
+        sakId: UUID,
+        periode: Periode,
+        saksbehandler: NavIdentBruker.Saksbehandler
+    ): Either<KunneIkkeRevurdere, Revurdering> {
+        if (!periode.erPeriodenIMånederEtter()) return KunneIkkeRevurdere.KanIkkeRevurdereInneværendeMånedEllerTidligere.left()
+
+        return hentSak(sakId)
+            .map { sak ->
+                val tilRevurdering = sak.behandlinger()
+                    .filter { it.status() == Behandling.BehandlingsStatus.IVERKSATT_INNVILGET }
+                    .filter { it.beregning()!!.getPeriode() inneholder periode }
+
+                if (tilRevurdering.isEmpty()) return KunneIkkeRevurdere.FantIngentingSomKanRevurderes.left()
+                if (tilRevurdering.size > 1) return KunneIkkeRevurdere.KanIkkeRevurderePerioderMedFlereAktiveStønadsperioder.left()
+
+                return tilRevurdering.single().let {
+                    val revurdering = OpprettetRevurdering(
+                        periode = periode,
+                        tilRevurdering = it,
+                        saksbehandler = saksbehandler
+                    )
+                    revurderingRepo.lagre(revurdering)
+                    revurdering.right()
+                }
+            }
+    }
+
+    override fun beregnOgSimuler(
+        revurderingId: UUID,
+        saksbehandler: NavIdentBruker.Saksbehandler,
+        fradrag: List<Fradrag>
+    ): Either<KunneIkkeRevurdere, SimulertRevurdering> {
+        return when (val revurdering = revurderingRepo.hent(revurderingId)!!) {
+            is BeregnetRevurdering, is OpprettetRevurdering, is SimulertRevurdering -> {
```

Comment:
I andre tilfelelr har vi en form for tilstandsmaskin som tar i mot en funksjon som gjør denne is avgjørelsen.